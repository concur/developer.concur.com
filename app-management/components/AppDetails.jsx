/* eslint-env browser */

import React, { PropTypes } from 'react';
import { Link } from 'react-router';

class AppDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      app: {},
      loading: false,
      error: false,
    };
  }

  componentWillMount() {
    window.fetch(`${process.env.API_SERVER}/apps/${this.props.params.id}`)
      .then(response => response.json())
      .then(app => this.setState({
        app,
        loading: false,
      }))
      .catch(() => this.setState({
        error: true,
        loading: false,
      }));
  }

  render () {
    if (this.state.loading) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>Loading...</p>
          </div>
        </div>
      );
    }

    if (this.state.error) {
      return (
        <div className="row">
          <div className="col-md-12">
            <p>An error occurred when loading this app</p>
            <Link to="/">Back to My Apps</Link>
          </div>
        </div>
      );
    }

    const { app } = this.state;

    return (
      <div className="row">
        <div className="col-md-12">
          <h2>{app.name}</h2>
          <Link to="/">&larr; Back to My Apps</Link>
          <div className="row">
            <p>{app.description}</p>
          </div>
        </div>
      </div>
    );
  }
}

AppDetails.propTypes = {
  params: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }),
};

export default AppDetails;
