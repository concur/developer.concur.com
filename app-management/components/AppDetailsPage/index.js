import { connect } from 'react-redux';
import AppDetailsPage from './AppDetailsPage';
import { fetchAppDetails, updateAppDetails } from '../../actions/appDetails';

export function mapStateToProps(state) {
  return state.appDetails;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppDetails: id => dispatch(fetchAppDetails(id)),
    handleSubmit: app => dispatch(updateAppDetails(app)),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppDetailsPage);
