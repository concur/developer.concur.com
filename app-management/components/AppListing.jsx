import React from 'react';
import { Link } from 'react-router';

class AppListing extends React.Component {
  constructor(props) {
    super(props);
    const apps = [
      {
        id: '2342asdfas',
        name: 'Starbucks',
      },
      {
        id: '234asdfasdf',
        name: 'Uber',
      },
      {
        id: 'asldfkajsdflj',
        name: 'My App',
      },
    ];
    this.state = { apps };
  }

  render () {
    const appItems = this.state.apps.map((app) => {
      const { id, name } = app;
      return (
        <div className="well" key={id}>
          <h2>{name}</h2>
          <Link to="" className="btn-u btn-u-green">Details</Link>
          <Link to="" className="btn-u btn-u-blue">Edit</Link>
        </div>
      );
    });

    return (
      <div className="row">
        <div className="col-md-12">
          <h1>My Apps</h1>
          <div className="row">
            {appItems}
          </div>
        </div>
      </div>
    );
  }
}

export default AppListing;
