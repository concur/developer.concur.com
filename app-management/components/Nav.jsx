import React from 'react';
import { Link } from 'react-router';

class Nav extends React.Component {
  render() {
    return (
      <div className="clearfix">
        <ul className="nav navbar-nav">
          <li><Link to={'/'} activeClassName="active">My Apps</Link></li>
          <li><Link to={'/new'} activeClassName="active">New App</Link></li>
          <li><Link to={'/signup'} activeClassName="active">Sign Up</Link></li>
          <li><Link to={'/login'} activeClassName="active">Login</Link></li>
          <li><Link to={'/logout'} activeClassName="active">Logout</Link></li>
        </ul>
      </div>
    );
  }
}

export default Nav;
