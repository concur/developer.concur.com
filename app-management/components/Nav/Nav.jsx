import React, { PropTypes } from 'react';
import { Link } from 'react-router';

const Nav = ({ authenticated }) => {
  const links = authenticated ? (
    <ul className="nav navbar-nav">
      <li>
        <Link to="/" activeClassName="active" onlyActiveOnIndex>
          <i className="fa fa-list" aria-hidden="true" /> My Apps
        </Link>
      </li>
      <li>
        <Link to="/new" activeClassName="active">
          <i className="fa fa-plus" aria-hidden="true" /> Create an App
        </Link>
      </li>
      <li>
        <Link to="/logout" activeClassName="active">
          <i className="fa fa-sign-out" aria-hidden="true" /> Logout
        </Link>
      </li>
    </ul>
  ) : (
    <ul className="nav navbar-nav">
      <li>
        <Link to="/signup" activeClassName="active">
          <i className="fa fa-user" aria-hidden="true" /> Sign Up
        </Link>
      </li>
      <li>
        <Link to="/login" activeClassName="active">
          <i className="fa fa-sign-in" aria-hidden="true" /> Login
        </Link>
      </li>
    </ul>
  );

  return (
    <div className="clearfix">
      {links}
    </div>
  );
};

Nav.propTypes = {
  authenticated: PropTypes.bool.isRequired,
};

export default Nav;
