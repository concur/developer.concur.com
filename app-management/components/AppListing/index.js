import { connect } from 'react-redux';
import AppListing from './AppListing';
import { fetchAppListing } from '../../actions/appListing';

export function mapStateToProps(state) {
  return state.appListing;
}

export function mapDispatchToProps(dispatch) {
  return {
    fetchAppListing: () => dispatch(fetchAppListing()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppListing);
