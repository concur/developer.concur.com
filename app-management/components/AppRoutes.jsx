import React, { PropTypes } from 'react';
import { Router, Route, IndexRoute, hashHistory } from 'react-router';

import PageContainer from './PageContainer';
import AppListing from './AppListing';
import AppDetailsPage from './AppDetailsPage';
import NewAppPage from './NewAppPage';
import LoginPage from './LoginPage';
import SignupPage from './SignupPage';
import LogoutPage from './LogoutPage';
import NotFound from './NotFound';

import { logout } from '../actions/auth';

class AppRoutes extends React.Component {
  constructor(props) {
    super(props);

    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.requireAuth = this.requireAuth.bind(this);
    this.showAuthForms = this.showAuthForms.bind(this);
    this.logout = this.logout.bind(this);
  }

  isAuthenticated() {
    const { store } = this.props;
    return !!store.getState().auth.token;
  }

  requireAuth(nextState, replaceState) {
    if (!this.isAuthenticated()) {
      replaceState('/login');
    }
  }

  showAuthForms(nextState, replaceState) {
    if (this.isAuthenticated()) {
      replaceState('/');
    }
  }

  logout(nextState, replaceState) {
    if (this.isAuthenticated()) {
      this.props.store.dispatch(logout());
    } else {
      replaceState('/login');
    }
  }

  render () {
    return (
      <Router history={hashHistory}>
        <Route path="/" component={PageContainer}>
          <IndexRoute component={AppListing} onEnter={this.requireAuth} />
          <Route path="new" component={NewAppPage} onEnter={this.requireAuth} />
          <Route path="details/:id" component={AppDetailsPage} onEnter={this.requireAuth} />
          <Route path="login" component={LoginPage} onEnter={this.showAuthForms} />
          <Route path="signup" component={SignupPage} onEnter={this.showAuthForms} />
          <Route path="logout" component={LogoutPage} onEnter={this.logout} />
          <Route path="*" component={NotFound} />
        </Route>
      </Router>
    );
  }
}

AppRoutes.propTypes = {
  store: PropTypes.shape({
    dispatch: PropTypes.func.isRequired,
  }),
};

export default AppRoutes;
