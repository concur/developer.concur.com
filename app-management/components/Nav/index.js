import { connect } from 'react-redux';
import Nav from './Nav';

export function mapStateToProps(state) {
  return {
    authenticated: !!state.auth.token,
  };
}

export default connect(
  mapStateToProps
)(Nav);
