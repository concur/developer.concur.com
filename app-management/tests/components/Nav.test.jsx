import React from 'react';
import { shallow } from 'enzyme';
import { Link } from 'react-router';
import { createStore } from 'redux';
import { shallowToJson } from 'enzyme-to-json';

import Nav from '../../components/Nav/Nav';
import { mapStateToProps } from '../../components/Nav';
import appReducer from '../../reducers';

const store = createStore(appReducer);

describe('<Nav />', () => {
  it('renders signup and login links when the user is not authenticated', () => {
    const page = shallow(<Nav authenticated={false} />);

    expect(shallowToJson(page)).toMatchSnapshot();
  });

  it('renders all other links when authenticated', () => {
    const page = shallow(<Nav authenticated={true} />);

    expect(shallowToJson(page)).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('returns authenticated state', () => {
      const state = mapStateToProps(store.getState());

      expect(state.authenticated).toBeDefined();
    });
  });
});
