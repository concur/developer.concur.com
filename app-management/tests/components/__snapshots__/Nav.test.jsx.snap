exports[`<Nav /> renders all other links when authenticated 1`] = `
<div
  className="clearfix">
  <ul
    className="nav navbar-nav">
    <li>
      <Link
        activeClassName="active"
        onlyActiveOnIndex={true}
        style={Object {}}
        to="/">
        <i
          aria-hidden="true"
          className="fa fa-list" />
         My Apps
      </Link>
    </li>
    <li>
      <Link
        activeClassName="active"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/new">
        <i
          aria-hidden="true"
          className="fa fa-plus" />
         Create an App
      </Link>
    </li>
    <li>
      <Link
        activeClassName="active"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/logout">
        <i
          aria-hidden="true"
          className="fa fa-sign-out" />
         Logout
      </Link>
    </li>
  </ul>
</div>
`;

exports[`<Nav /> renders signup and login links when the user is not authenticated 1`] = `
<div
  className="clearfix">
  <ul
    className="nav navbar-nav">
    <li>
      <Link
        activeClassName="active"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/signup">
        <i
          aria-hidden="true"
          className="fa fa-user" />
         Sign Up
      </Link>
    </li>
    <li>
      <Link
        activeClassName="active"
        onlyActiveOnIndex={false}
        style={Object {}}
        to="/login">
        <i
          aria-hidden="true"
          className="fa fa-sign-in" />
         Login
      </Link>
    </li>
  </ul>
</div>
`;
